### DISCUSSION TITLE
The transforms.SaveImaged saved nii segmentation mask have different origin and direction with original data

### DISCUSSION THREAD

**Discussion by auroua** [Project-MONAI/MONAI#4051] [Q&A]
As shown in the following image, the `transforms.SaveImaged` saved nii segmentation mask have different origin and direction with original data.
![Screenshot_2022-04-02_16-29-36](https://user-images.githubusercontent.com/2869412/161374584-8ede1915-b137-4fac-9f76-3400e3a70c0d.png)

I also used `SimpleITK` to read the origin and direction of the saved segmetntation mask, the values are different with the original `nii` data. How to keep the `transformer.SaveImaged` saved data have the same origin and direction with the original `nii` data.
Code for reading the origin and direction of `nii` data.
```
    original_itk_data = sitk.ReadImage(test_data_path)
    original_label_data = sitk.ReadImage(test_label_path)

    original_spacing = original_itk_data.GetSpacing()
    original_origin = original_itk_data.GetOrigin()
    original_direction = original_itk_data.GetDirection()
    print(original_spacing, original_origin, original_direction)

    pred_spacing = original_label_data.GetSpacing()
    pred_origin = original_label_data.GetOrigin()
    pred_direction = original_label_data.GetDirection()
    print(pred_spacing, pred_origin, pred_direction)
```

The output is :
```
(0.68359375, 0.68359375, 3.0) (0.0, 0.0, 0.0) (1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0)
(0.68359375, 0.68359375, 3.0) (-174.658203125, 174.658203125, -205.5) (1.0, 0.0, 0.0, 0.0, -1.0, 0.0, 0.0, 0.0, 1.0)
```

**Comment by Nic-Ma** [comment]
Hi @auroua ,

Thanks for your feedback.
Could you please help share your full program or at least transform chain for further analysis?
I think you can refer to below example to invert your transforms to align with original image:
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/torch/unet_inference_dict.py#L70

@wyli I think maybe we can make a category / milestone to track all the metadata-related issues and discussions?
I feel we already got much feedback about metadata, then we can ask the users to help review when the metadata feature is ready?

Thanks in advance.

**Reply by auroua** [reply]
Hi @Nic-Ma, the following is the code used to generate the output masks.
```
    def visualize_3d(self, cfgs):
        datalist_json = os.path.join(cfgs.DATASETS.DATA_DIR, cfgs.DATASETS.JSON)

        test_org_transforms = transforms.Compose(
            [
                transforms.LoadImaged(keys="image"),
                transforms.EnsureChannelFirstd(keys="image"),
                transforms.Orientationd(keys=["image"],
                                        axcodes=cfgs.INPUT.ORIENTATION),
                transforms.Spacingd(keys=["image"],
                                    pixdim=(cfgs.INPUT.SPACING.X, cfgs.INPUT.SPACING.Y, cfgs.INPUT.SPACING.Z),
                                    mode="bilinear"),
                transforms.ScaleIntensityRanged(keys=["image"],
                                                a_min=cfgs.INPUT.SCALE_INTENSITY.ORIGINAL_MIN,
                                                a_max=cfgs.INPUT.SCALE_INTENSITY.ORIGINAL_MAX,
                                                b_min=cfgs.INPUT.SCALE_INTENSITY.TARGET_MIN,
                                                b_max=cfgs.INPUT.SCALE_INTENSITY.TARGET_MAX,
                                                clip=True),
                transforms.CropForegroundd(keys=["image"], source_key="image"),
                transforms.EnsureTyped(keys="image"),
            ]
        )
        test_files = load_decathlon_datalist(datalist_json,
                                             True,
                                             cfgs.DATASETS.TEST_TYPE,
                                             base_dir=cfgs.DATASETS.DATA_DIR)
        test_ds = data.Dataset(data=test_files, transform=test_org_transforms)
        test_loader = data.DataLoader(test_ds,
                                      batch_size=1,
                                      shuffle=False,
                                      num_workers=cfgs.DATALOADER.TEST_WORKERS,
                                      sampler=None,
                                      pin_memory=True,
                                      persistent_workers=True)

        post_transforms = transforms.Compose([
            transforms.EnsureTyped(keys="pred"),
            transforms.Invertd(
                keys="pred",
                transform=test_org_transforms,
                orig_keys="image",
                meta_keys="pred_meta_dict",
                orig_meta_keys="image_meta_dict",
                meta_key_postfix="meta_dict",
                nearest_interp=False,
                to_tensor=True,
            ),
            transforms.AsDiscreted(keys="pred", argmax=True, to_onehot=5),
            transforms.SplitChanneld(keys="pred", channel_dim=0),
            transforms.SaveImaged(keys="pred_0",
                                  meta_keys="pred_meta_dict",
                                  output_dir=os.path.join(cfgs.OUTPUT_DIR, "inference_all_part1"),
                                  output_postfix="seg_0",
                                  resample=False),
            transforms.SaveImaged(keys="pred_1",
                                  meta_keys="pred_meta_dict",
                                  output_dir=os.path.join(cfgs.OUTPUT_DIR, "inference_all_part1"),
                                  output_postfix="seg_1",
                                  resample=False),
            transforms.SaveImaged(keys="pred_2",
                                  meta_keys="pred_meta_dict",
                                  output_dir=os.path.join(cfgs.OUTPUT_DIR, "inference_all_part1"),
                                  output_postfix="seg_2",
                                  resample=False),
            transforms.SaveImaged(keys="pred_3",
                                  meta_keys="pred_meta_dict",
                                  output_dir=os.path.join(cfgs.OUTPUT_DIR, "inference_all_part1"),
                                  output_postfix="seg_3",
                                  resample=False),
            transforms.SaveImaged(keys="pred_4",
                                  meta_keys="pred_meta_dict",
                                  output_dir=os.path.join(cfgs.OUTPUT_DIR, "inference_all_part1"),
                                  output_postfix="seg_4",
                                  resample=False)
        ])

        loader = LoadImage()

        with torch.no_grad():
            for test_data in test_loader:
                test_data["image"] = test_data["image"].cuda()
                img_name = test_data['image_meta_dict']['filename_or_obj'][0].split('/')[-1]
                self.logger.info("Inference on case {}".format(img_name))
                test_data["pred"] = sliding_window_inference(test_data["image"],
                                                             roi_size=(self.cfg.INPUT.RAND_CROP.ROI.X,
                                                                       self.cfg.INPUT.RAND_CROP.ROI.Y,
                                                                       self.cfg.INPUT.RAND_CROP.ROI.Z),
                                                             sw_batch_size=self.cfg.SOLVER.SW_BATCH_SIZE_TEST,
                                                             predictor=self.model,
                                                             overlap=self.cfg.SOLVER.INFER_OVERLAP)
                test_data = [post_transforms(i) for i in decollate_batch(test_data)]
```

**Reply by Nic-Ma** [reply]
Hi @wyli ,

As we just totally updated the image writing logic recently, maybe there is some compatible issue with inverting?

Thanks.

**Comment by wyli** [comment]
I didn't look into the details but it looks like the same issue as https://github.com/Project-MONAI/tutorials/discussions/481. I put a patch here https://github.com/Project-MONAI/MONAI/pull/3882. I think we should merge #3882 for now and change it if we find a better default.

**Comment by adwaykanhere** [comment]
Hi, I'm still facing the same issue that @auroua mentioned. Please suggest how to fix it. I'm using the following script:

```
post_transforms_fl_liver = Compose([
    Invertd(
        keys="pred",
        transform=test_org_transforms,
        orig_keys="image",
        meta_keys="pred_meta_dict",
        orig_meta_keys="image_meta_dict",
        meta_key_postfix="meta_dict",
        nearest_interp=False,
        to_tensor=True,
        device="cpu",
    ),
    AsDiscreted(keys="pred", argmax = True),
    Orientationd(keys=["pred"], axcodes="RPI"),
    EnsureChannelFirstd(keys="image"),
    Spacingd(keys=["pred"], pixdim=(
            1.5, 1.5, 2.0), mode="bilinear"),    
    SaveImaged(keys="pred", meta_keys="image_meta_dict", output_dir="/content/s_all_outputs", output_postfix="f1_liver", resample=True),
])

model_liver.load_state_dict(torch.load(
    os.path.join(root_dir, "best_metric_model_liver_128.pth")))
model_liver.eval()

with torch.no_grad():
    for test_data in test_org_loader:
        test_inputs = test_data["image"].to(device)
        roi_size = (160, 160, 160)
        sw_batch_size = 4
        test_data["pred"] = sliding_window_inference(
            test_inputs, roi_size, sw_batch_size, model_liver)

        test_data = [post_transforms_fl_liver(i) for i in decollate_batch(test_data)]
```

**Reply by KumoLiu** [reply]
Hi @adwaykanhere, I noticed that you added some transforms after the `invertd`, if `test_org_transforms` have included these, it should be unnecessary to add, you can refer to [this tutorial](https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/torch/unet_inference_dict.py) to modify your `post_transforms_fl_liver `. Also if your monai version is not the latest, maybe you can try to update to 1.1.0 and test again.
https://github.com/Project-MONAI/tutorials/blob/7403995f12f4198042eb1a52f3d2e1c88d53ae71/3d_segmentation/torch/unet_inference_dict.py#L65-L79

Hope it can help you, thanks!

**Reply by adwaykanhere** [reply]
Hi @KumoLiu I tried your suggestions and I'm still getting the same error.

**Reply by KumoLiu** [reply]
Hi @adwaykanhere, could you please share more details then I could reproduce the issue? Thanks!
